bokeh/sphinxext/bokeh_sampledata_xref.py

```diff
@@ -109,50 +128,74 @@ def process_sampledata_xrefs(app, doctree, fromdocname):
-        sampledata_refs = []
-        refuris = []
-        for s in env.all_sampledata_xrefs:
-            refuri = app.builder.get_relative_uri(
-                fromdocname, s['docname']
-            )
-            if s["keyword"] == node.sampledata_key and "gallery" in refuri:
-                sampledata_refs.append(s)
-                refuris.append(refuri)
-
-        _len = len(sampledata_refs)
-        para = nodes.paragraph()
-        if _len:
-            s = '' if _len==1 else 's'
-            description = (
-                _(f'See the following example{s} that use this sample data set: ')
-            )
-        else:
-            description = (_('There are no references for this sample data set'))
-        para += nodes.Text(description, description)
-        for i, (sample_info, refuri) in enumerate(zip(sampledata_refs, refuris)):
-            # Create references
-            newnode = nodes.reference('', '')
-            ref_name = basename(sample_info['docname'])
-            innernode = nodes.emphasis(_(ref_name), _(ref_name))
-            newnode['refdocname'] = sample_info['docname']
-            newnode['refuri'] = refuri
-            newnode.append(innernode)
-            para += newnode
-            if 1<_len:
-                _l =  _len-2
-                if i == _l:
-                    para += nodes.Text(' and ', ' and ')
-                elif i < _l:
-                    para += nodes.Text(', ', ', ')
-
-        para += nodes.Text('.', '.')
-        content.append(para)
+def purge_gallery_xrefs(app, env, docname):
+    if not hasattr(env, 'all_gallery_overview'):
+        return
+
+    env.all_gallery_overview = [
+        xref for xref in env.all_gallery_overview if xref['docname'] != docname
+    ]
+
+def merge_gallery_xrefs(app, env, docnames, other):
+    if not hasattr(env, 'all_gallery_overview'):
+        env.all_gallery_overview = []
+
+    if hasattr(other, 'all_gallery_overview'):
+        env.all_gallery_overview.extend(other.all_gallery_overview)
+
+def process_gallery_overview(app, doctree, fromdocname):
+
+    env = app.builder.env
+
+    if not hasattr(env, 'all_gallery_overview'):
+        env.all_gallery_overview = []
+
+    for node in doctree.traverse(gallery_xrefs):
+
+        ref_dict = {}
+        for s in env.all_gallery_overview:
+            sp = s['docname'].split('/')
+            if node.subfolder == 'all' or sp[-2] == node.subfolder:
+                letter = sp[-1][0].upper()
+                if letter in ref_dict:
+                    ref_dict[letter].append(s)
+                else:
+                    ref_dict[letter] = [s]
+
+        content = []
+        for letter, refs in ref_dict.items():
+            para = nodes.paragraph()
+            para += nodes.rubric((_(letter)), (_(letter)))
+            for ref in refs:
+                ref_name = basename(ref['docname'])
+                if node.subfolder == 'all':
+                    ref_name += f" ({ref['docname'].split('/')[-2]})"
```

Comment:
This is to add `(plotting)` or `(models)` to the link. I do not like it. IMHO this it to much and does imrpove UX.